internal/migrate/versions.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package migrate
+
+import (
+	"database/sql"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+)
+
+/*
+
+Scenarios it doesn't handle properly.
+
+1. Rollback to initial state on multi-step migration.
+
+	Let's say there's a scenario where we run migration steps:
+	1. update a table schema
+	2. move files
+	3. update a table schema
+	4. update a table schema, which fails
+
+	In this case there's no easy way to rollback the moving of files.
+
+2. Undoing migrations.
+
+	Intentionally left out, because we do not gain that much from currently.
+
+3. Snapshotting the whole state.
+
+	This probably should be done by the user of this library, when there's disk-space available.
+
+4. Figuring out what the exact executed steps are.
+*/
+
+// Migration describes a migration steps
+type Migration struct {
+	Table    string
+	OnCreate Action
+	Steps    []*Step
+}
+
+// Step describes a single step in migration.
+type Step struct {
+	Description string
+	Version     int // Versions should start at 1
+	Action      Action
+}
+
+// Action is something that needs to be done
+type Action interface {
+	Run(log *zap.Logger, db DB, tx *sql.Tx) error
+}
+
+// ValidTableName checks whether the specified table name is valid
+func (migration *Migration) ValidTableName() error {
+	matched, err := regexp.MatchString(`^[a-z_]+$`, migration.Table)
+	if !matched || err != nil {
+		return Error.New("invalid table name: %v", migration.Table)
+	}
+	return nil
+}
+
+// Run runs the migration steps
+func (migration *Migration) Run(log *zap.Logger, db DB) error {
+	err := migration.ValidTableName()
+	if err != nil {
+		return err
+	}
+
+	err = migration.createVersionTable(log, db)
+	if err != nil {
+		return Error.New("creating version table failed: %v", err)
+	}
+
+	version, err := migration.getLatestVersion(log, db)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if version < 0 {
+		log.Info("Latest Version", zap.Int("version", version))
+	} else {
+		log.Info("No Version")
+	}
+
+	if version < 0 && migration.OnCreate != nil {
+		log := log.Named("init")
+		tx, err := db.Begin()
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = migration.OnCreate.Run(log, db, tx)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		err = migration.addVersion(tx, db, 0)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		if err := tx.Commit(); err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	for _, step := range migration.Steps {
```

Comment:
nit: this could just be a standard for x=, x<; x++ loop without the continue

Reply:
With versioned migrations I've seen steps removed/optimized eventually, so would prefer with continue... it doesn't pose a performance concern either.